### PR TITLE
AnonCreds Credentials using the W3C Standard - wrappers

### PR DESCRIPTION
**Pull request by Artemkaaas**
* Updated Python wrapper to expose new methods
* Update NodeJS and React Native wrappers to expose new methods

### CODE REVIEW

File: src/data_types/credential.rs
Comment (gvelez17) [inline src/data_types/credential.rs:104]: just curious what is the reason for change from Short to Raw is this because we need richer data for the flex creds?

Comment (Artemkaaas) [reply src/data_types/credential.rs:104]: I just considered name `short` not clear and confusing. can be reverted

File: wrappers/python/anoncreds/bindings.py
Comment (gvelez17) [inline wrappers/python/anoncreds/bindings.py:1050]: @Artemkaaas why change the signature of the attributes?  the original create_credential is

```
 attr_raw_values: Mapping[str, str],
    attr_enc_values: Optional[Mapping[str, str]],
    revocation_config: Optional[RevocationConfig],
    ```
    
    maybe good to keep the signature the same if we can?

Comment (Artemkaaas) [reply wrappers/python/anoncreds/bindings.py:1050]: It was decided to move encoding of credential values to the library

File: wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts
Comment (TimoGlastra) [inline wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts:105]: Is there a reason why there is a separate method for w3c? Getting an attribute is the same for w3c and non w3c credentials right?

File: wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts
Comment (TimoGlastra) [inline wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts:630]:
```suggestion
  public createW3cCredentialOffer(options: {
```

To not have confusion in casing, we always only capitalize the first letter of a word (and W3c is one word in this case)

File: wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts
Comment (TimoGlastra) [inline wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts:638]: As I mentioned in another comment before, I do't think we need a w3c credential offer method, as the offer is not any different based on whether it's a w3c or anoncreds offer? 

The only difference is the end-result: whether the credential is encoded as AnonCreds or W3C credential

File: wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts
Comment (TimoGlastra) [inline wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts:649]: Same here: `W3c`

and also, I don't see a reason for a w3c credential request, only the end result is different I think?

File: wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts
Comment (TimoGlastra) [inline wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts:783]: I think there's a lot of code here for w3c methods that is basically the same except for the end result. Do you think we can reuse some code here? I think this can lead to inconsistencies in functionality

File: wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts
Comment (TimoGlastra) [inline wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts:946]: As mentioned in other PR, I don't think this should be handled by the AnonCreds library

File: wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts
Comment (TimoGlastra) [inline wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts:932]: This method is basically duplicate from the method below

File: wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts
Comment (TimoGlastra) [inline wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts:980]: This feels like it's not really useful in the context of AnonCreds credentials, as you can't use them. Honestly, I'd like to keep it simple and not expose these kind of methods, but just have a simple method like you have with the normal anoncreds encoding (sign based on the attribute map you pass).

That way you can get an anoncreds credential in w3c encoding, if you want to add other contexts, or do special things for using it with other proof types, that should be handled outside the anoncreds library. It has nothing to do with AnonCreds. IMO this library should do the minimal amount of processing it can, and all custom ways should be handled by the frameworks consuming this library.

File: wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts
Comment (TimoGlastra) [inline wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts:1099]: As these just call `anoncreds_credential_offer_from_json` I think we shouldn't add these new methods. Duplicating all methods doesn't make sense to me, and I think we should keep the duplication of methods only to the methods where it's actually needed

File: wrappers/javascript/packages/anoncreds-nodejs/tests/api.test.ts
Comment (TimoGlastra) [inline wrappers/javascript/packages/anoncreds-nodejs/tests/api.test.ts:611]: This can just be `CredentialOffer.create` right?

File: wrappers/javascript/packages/anoncreds-nodejs/tests/api.test.ts
Comment (TimoGlastra) [inline wrappers/javascript/packages/anoncreds-nodejs/tests/api.test.ts:620]: Same here

```suggestion
    const { credentialRequestMetadata, credentialRequest } = CredentialRequest.create({
```

File: wrappers/javascript/packages/anoncreds-nodejs/tests/api.test.ts
Comment (TimoGlastra) [inline wrappers/javascript/packages/anoncreds-nodejs/tests/api.test.ts:628]:
```suggestion
    const credential = W3cCredential.create({
```

File: wrappers/javascript/packages/anoncreds-react-native/src/ReactNativeAnoncreds.ts
Comment (TimoGlastra) [inline wrappers/javascript/packages/anoncreds-react-native/src/ReactNativeAnoncreds.ts:414]: Would it be possible to fix these @ts-ignores? 

It will disable any type checking for this line, and can result in type issues down the line

File: wrappers/javascript/packages/anoncreds-shared/src/api/Credential.ts
Comment (TimoGlastra) [inline wrappers/javascript/packages/anoncreds-shared/src/api/Credential.ts:176]: Won't comment again, but would like to use `w3c` or `W3c` (depending on whether it's the start of the variable /method name), but not use `W3C` (all caps)

File: wrappers/javascript/packages/anoncreds-shared/src/api/W3CCredential.ts
Comment (TimoGlastra) [inline wrappers/javascript/packages/anoncreds-shared/src/api/W3CCredential.ts:225]: In JS it's fine to use `type` as a variable name I think. Is this taken over from the Python wrapper?

Would prefer to just use `type`, or otherwise if you want to avoid using `type`, use something like `credentialType`

**Review by TimoGlastra — CHANGES_REQUESTED** [review]
Mostly looked at the JS wrapper ,as I have more experience with it. 

I think overall looks good, but I think we can reduce the number of method exposed in the wrapper, as well as have some more consistency in using w3c in names

**Comment by Artemkaaas** [comment]
@TimoGlastra I processed and resolved comments connected to using `W3c` instead `W3C`. 
I will handle the rest of comments (relating to credential-offer/request duplication and credential methods for non-anoncreds proof) after discuss at the WG call. Unfortunately, I may miss today's call due to some personal errands.

File: wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts
Comment (Artemkaaas) [reply wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts:105]: I do not understand your point. `w3cCredentialGetAttribute` method calls `anoncreds_w3c_credential_get_attribute` external Rust function which obtain W3C credential object from handle and returns value for requesting attribute.

Comment (TimoGlastra) [reply wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts:105]: Yeah I was thinking more of that whether the object is an AnonCreds representation or a W3C representation doesn't really change the input or output of this method, so from an user POV it could be the same method right?

File: wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts
Comment (Artemkaaas) [reply wrappers/javascript/packages/anoncreds-nodejs/src/NodeJSAnoncreds.ts]: I added helper methods doing conversion of parameters in order to reduce code duplication. 
But left w3c* methods. 
I think methods here should be mapped one to one with external C API. Do not see a sense to combine w3c and legacy methods.

**Comment by TimoGlastra** [comment]
@Artemkaaas once #291 is merged, I'll take another look at this PR, but there's currently too much different changes

**Comment by swcurran** [comment]
The plan (per requests from @TimoGlastra and @andrewwhitehead ) is to merge #273 first (once the test failure is fixed), so this is easier to check. Tests are passing here, so hopefully this one will be easy to approve once #273 is merged.

FYI @Artemkaaas 

Thanks

**Comment by swcurran** [comment]
@TimoGlastra , @andrewwhitehead -- can you please do a review of this PR. The other two have been merged, so this should be easier to review.  Thanks.

File: docs/design/w3c/w3c-representation.md
Comment (andrewwhitehead) [inline docs/design/w3c/w3c-representation.md:546]: Do we not want to keep the examples?

File: wrappers/javascript/packages/anoncreds-shared/src/api/index.ts
Comment (andrewwhitehead) [inline wrappers/javascript/packages/anoncreds-shared/src/api/index.ts:19]: CredentialOffer is already exported

File: wrappers/python/anoncreds/types.py
Comment (andrewwhitehead) [inline wrappers/python/anoncreds/types.py:448]: Comparisons to None should use 'is'

**Comment by andrewwhitehead** [comment]
Looks good, with a couple tiny notes. We'll want to add the w3c tests to the CI.

**Review by TimoGlastra — APPROVED** [review]
LGTM! I already played with the api a bit for creating the test vectors and it made sense to me

Only thing is maybe that there's quite some duplication between the w3c and non w3c code in the JS wrapper, but we can look at that in the future maybe. Would ne good to have this merged
And reelased!

**Comment by TimoGlastra** [comment]
Also one more question/
: did you test the react native code already? Or should we do some
Testing in react native?

**Comment by Artemkaaas** [comment]
> Also one more question/ : did you test the react native code already? Or should we do some Testing in react native?

I only checked build and compile scripts for wrapper, but not tested integration into a react-native application.

**Comment by TimoGlastra** [comment]
> > Also one more question/ : did you test the react native code already? Or should we do some Testing in react native?
> 
> I only checked build and compile scripts for wrapper, but not tested integration into a react-native application.

We can do that once this is merged and released as dev 👍

**Pull request merged**
